### PR TITLE
feat(server): custom library scanning interval

### PR DESCRIPTION
**Pull request by danieldietzler**
Screenshot 2023-10-26 at 19 50 49

### CODE REVIEW

File: server/package-lock.json
Comment (jrasm91) [inline server/package-lock.json:1]: This upgrades the package lock from version 2 to 3. Can you revert this change, then re-install the new dependency? It should add it without upgrading the lockfile.

File: server/src/domain/library/library.service.ts [hunk 1]
```diff
@@ -33,17 +36,45 @@ import {
...
+  }
+
+  async init() {
+    const config = await this.configCore.getConfig();
+    const libraryScanJob = new CronJob(
+      config.libraryScan.cronExpression,
+      async () => {
+        await this.jobRepository.queue({ name: JobName.LIBRARY_QUEUE_SCAN_ALL, data: { force: false } });
+      },
+      undefined,
```
Comment (alextran1502) [inline server/src/domain/library/library.service.ts:62]: Can we add the comments here for each of the `undefined` argument to know which parameter they are represented?

File: server/src/domain/job/job.service.ts [hunk 2]
```diff
@@ -28,6 +29,7 @@ export class JobService {
     @Inject(IPersonRepository) private personRepository: IPersonRepository,
   ) {
     this.configCore = SystemConfigCore.create(configRepository);
+    this.configCore.addValidator((config) => this.validateCronExpression(config.library.scan.cronExpression));
```
Comment (jrasm91) [inline server/src/domain/job/job.service.ts:32]: The cron expression is for the library service, so it should be the one subscribing to updates and throwing an error for an invalid configuration.

File: server/src/domain/job/job.service.ts [hunk 3]
```diff
@@ -228,4 +229,12 @@ export class JobService {
         break;
     }
   }
+
+  private validateCronExpression(cronExpression: string) {
```
Comment (jrasm91) [inline server/src/domain/job/job.service.ts:233]: I think we need to move this into a repository. We don't need to be referencing importing/using `cron` directly in the domain.

File: server/src/domain/library/library.service.ts [hunk 4]
```diff
@@ -33,17 +36,49 @@ import {
...
+      // runOnInit
+      undefined,
+      // utcOffset
+      undefined,
+      // prevents memory leaking by automatically stopping when the node process finishes
+      true,
+    );
+
+    this.configCore.config$.subscribe((config) => {
```
Comment (jrasm91) [inline server/src/domain/library/library.service.ts:78]: I see you subscribe to changes here, but the validation is elsewhere. They should be together.

File: server/src/domain/library/library.service.ts [hunk 5]
```diff
@@ -33,17 +36,49 @@ import {
...
+      libraryScanJob.setTime(new CronTime(config.library.scan.cronExpression));
```
Comment (jrasm91) [inline server/src/domain/library/library.service.ts:79]: nestjs has a default scheduler module, which is exactly for stuff like this. I think we should possibly add new methods to the job repository for job scheduling. So something like
```typescript
  addCronJob(name: string, expression: string);
  updateCronJob(name: string, expression: string);
  removeCronJob(name: string);
```

This way you can write unit tests against it and you don't care about what library or implementation is used behind the scenes. We should probably import the SchedulerRegistry in the infra/job.repository implementation though.

```typescript
constructor(private schedulerRegistry: SchedulerRegistry) {}
```

See https://docs.nestjs.com/techniques/task-scheduling#dynamic-cron-jobs

**Review by jrasm91 — COMMENTED** [review]
The system config changes look great, but I think we can improve the scheduling implementation. Specifically by moving the actual schedule management out of the domain/ and into the infra/ and probably converting the implementation to use @nestjs/schedule where possible.

**Comment by jrasm91** [comment]
I think it would be cool too, to have this work like the storage template, which has a dropdown of a bunch of default options/values, but ultimately just uses what is in the custom input field. Maybe we can add some options like (every night at midnight, every night at 1am, every 6 hours, etc.

**Comment by alextran1502** [comment]
Hi Daniel, is there any more work needed to be done in this PR or it is good to test and merge?

**Comment by danieldietzler** [comment]
> Hi Daniel, is there any more work needed to be done in this PR or it is good to test and merge?

Should be fine :)

**Comment by nodis** [comment]
This is so cool, exactly what I needed, thank you for your hard work

File: server/src/infra/repositories/job.repository.ts [hunk 6]
```diff
@@ -1,23 +1,80 @@
...
+    private schedulerReqistry: SchedulerRegistry,
+  ) {}
 
   addHandler(queueName: QueueName, concurrency: number, handler: (item: JobItem) => Promise<void>) {
     const workerHandler: Processor = async (job: Job) => handler(job as JobItem);
     const workerOptions: WorkerOptions = { ...bullConfig, concurrency };
     this.workers[queueName] = new Worker(queueName, workerHandler, workerOptions);
   }
 
+  validateCronExpression(expression: string) {
```
Comment (jrasm91) [inline server/src/infra/repositories/job.repository.ts:35]: If we are using the class-validator why is this needed still?

**Review by jrasm91 — APPROVED** [review]
LGTM more or less :smile:

File: server/src/infra/repositories/job.repository.ts (hunk 6 above)
Comment (danieldietzler) [reply server/src/infra/repositories/job.repository.ts:35]: Do validate e.g. config files

Comment (danieldietzler) [reply server/src/infra/repositories/job.repository.ts:35]: To validate e.g. config files and also just to make really sure that the expression is valid because otherwise the system will die quite badly.
